[GSoC] Higher order system of ODEs to first order reduction

mijo2 opened the pull request:
#### References to other Issues or PRs
Fixes #19238

#### Brief description of what is fixed or changed
Added functionalities in `systems.py` that can reduce higher-order ODEs to first-order ODEs and solve the reduced system.

Removes the following solvers:
`Linear, 2 equations, Order 2, Type 1-6, and Type 8-11.`

The solvers are removed by implementing the following:
1. Method to reduce any higher-order system into a first-order system by introducing dummy variables. This type of method doesn't always work to solve every type of higher-order system because of the fact that we can only solve a first-order system whose coefficient matrix is commutative with its antiderivative or which can be divided into systems having commutative antiderivative. This method was responsible for the `removal of Type 1-4 from the old second-order solvers` meanwhile, it also extended Sympy's capabilities to solve systems of ODEs.

2. Method to reduce a second-order system `X'' = A*(t*X' - X) + b(t)` to a solvable first-order system where `A` is a constant matrix, `X` is a vector of dependent variables(same nomenclature used below). This removed Types 5, 8, and 11.

3. Method to transform a second-order system `X'' = f(t)*A*X` where `f(t) = 1/(a*t**2 + b*t + c)**2` into a constant coefficient second-order system. This removed Types 6 and 10.

4. Method to solve standard Cauchy-Euler systems not just of second-order with two equations but of any order with any number of equations. This removed Type 9 solver.

5. Method to compute the solution for first-order systems of type: `X' = f(t)*A*X + b` where `f(t)` is any invertible function. This solution is added as Type 5(homogeneous) and Type 6(non-homogeneous) in `linodesolve`.

6. Method to reduce a system of type: `X(n)(t) = f(t) * A * X + b(t)` where `X(n)(t)` is the nth order derivative of the vector `X`, `A` is a constant matrix and `f(t)` is a scalar. This system is transformed into a system that can be divided into smaller sub-systems which can be divided and solved using strongly and weakly connected components logic. But not all the systems of this form can be handled by the module. Only some of them. More details about the component division logic given [here](https://github.com/sympy/sympy/wiki/ODE-Systems-roadmap).

```
Note: Work for implementing a function that can reduce a dependent variable when the minimum order of the dependent variables is greater than 0 and implementing a method to solve X'' = f(t)*A*X is still remaining and has to be carried out in future PR/s.
```

#### Other comments

Examples below show how `dsolve_system` is now able to solve many types of higher-order systems:

1. Using method 1 of introducing dummy variables to solve higher-order system:
```
>>> eqs = [Eq(f(x).diff(x, 2), 2*f(x) + g(x)), Eq(g(x).diff(x, 2), -f(x))]
>>> dsolve(eqs)
[Eq(f(x), C3*x*exp(x) + C3*exp(x) + C4*exp(x) + (-C1 - C2*x + C2)*exp(-x)),
Eq(g(x), -C3*x*exp(x) + C3*exp(x) - C4*exp(x) + (C1 + C2*x + C2)*exp(-x))]
```

2. Solving second-order systems of form `X'' = A*(t*X' - X) + b(t)`:
```
>>> eqs = [Eq(f(x).diff(x, 2), 2/x *(x*g(x).diff(x) - g(x))), Eq(g(x).diff(x, 2),-2/x *(x*f(x).diff(x) - f(x)))]
>>> dsolve(eqs)
[Eq(f(x), C1*x + x*Integral(C2*exp(-x_)*sin(2*exp(x_)) + C3*exp(-x_)*cos(2*exp(x_)), (x_, log(x)))),
Eq(g(x), C4*x + x*Integral(C2*exp(-x_)*cos(2*exp(x_)) - C3*exp(-x_)*sin(2*exp(x_)), (x_, log(x))))]
```

3. Solving second-order systems of the form `X'' = f(t)*A*X`` where `f(t)` is inverse of the square of a quadratic function in `t`:
```
>>> eqs = [Eq(Derivative(f(t), (t, 2)), f(t)/t**4), Eq(Derivative(g(t), (t, 2)), d*g(t)/t**4)]
>>> dsolve(eqs)
[Eq(f(t), (C1*exp(-1/t) - C2*exp(1/t))*sqrt(t**2)),
Eq(g(t), (C3*exp(-sqrt(d)/t)/sqrt(d) - C4*exp(sqrt(d)/t)/sqrt(d))*sqrt(t**2))]
```

4. Solving a Cauchy-Euler system:
```
>>> eqs = [Eq(Derivative(f(t), (t, 2)), t*f(t)), Eq(Derivative(g(t), (t, 2)), t*g(t))]
>>> dsolve(eqs)
[Eq(f(t), C1*t**(1/2 - sqrt(5)/2)*(-3/2 - sqrt(5)/2) + C2*t**(1/2 + sqrt(5)/2)*(-3/2 + sqrt(5)/2) + C3*t**(1 - sqrt(2))*(-sqrt(2) - 1) + C4*t**(1 + sqrt(2))*(-1 + sqrt(2))),
 Eq(g(t), C1*t**(1/2 - sqrt(5)/2)*(-sqrt(5)/2 - 1/2) + C2*t**(1/2 + sqrt(5)/2)*(-1/2 + sqrt(5)/2))]
```

5. Solving systems of the form `X' = f(t)*A*X + b(t)` was already possible, this method was added for faster computation.

6. Solving systems of the form: `X(n)(t) = f(t) *A*X + b(t)`:
```
>>> eqs 
[Eq(Derivative(f(t), (t, 2)), t*f(t) + t*g(t)),
 Eq(Derivative(g(t), (t, 2)), 2*t*f(t) + t*g(t))]

>>> dsolve(eqs)                                                                                                                                                                                 
[[Eq(f(t), -sqrt(2)*(C1*airyai(-t*(-1 + sqrt(2))**(1/3)) + C2*airybi(-t*(-1 + sqrt(2))**(1/3)))/2 + sqrt(2)*(C3*airyai(t*(1 + sqrt(2))**(1/3)) + C4*airybi(t*(1 + sqrt(2))**(1/3)))/2),
  Eq(g(t), C1*airyai(-t*(-1 + sqrt(2))**(1/3)) + C2*airybi(-t*(-1 + sqrt(2))**(1/3)) + C3*airyai(t*(1 + sqrt(2))**(1/3)) + C4*airybi(t*(1 + sqrt(2))**(1/3)))]]
```

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* solvers
  * Solving higher-order system of ODEs by reducing/transforming it into the first-order system of ODEs by numerous methods.
<!-- END RELEASE NOTES -->

mijo2 committed:
solvers.ode: Added function that reduces higherorder ODEs

Higher order ODEs can be reduced to first order ODEs using a
private function added in systems.py

mijo2 committed:
solvers.ode: Moved some of the test cases

mijo2 committed:
solvers.ode: Moved some more test cases and identified a fail case

mijo2 commented:
@oscarbenjamin I have found out a weird test case that works only in a particular order:

```
In [186]: eqs = [Eq(-9*k(t) + 7*f(t) + h(t).diff(t), 0), Eq(9*h(t) + 7*g(t) + k(t).diff(t), 0), Eq(f(t).diff(t), h(t)), Eq(g(t).diff(t), k(t))]                                                             

In [187]: sol = dsolve(eqs)                                                                                                                                                                                 

In [188]: checkodesol(eqs, sol)                                                                                                                                                                             
Out[188]: (True, [0, 0, 0, 0])
```

But, given a different order, the solution messes up:
```
In [196]: eqs                                                                                                                                                                                               
Out[196]: 
[Eq(Derivative(f(t), t), h(t)),
 Eq(Derivative(g(t), t), k(t)),
 Eq(Derivative(h(t), t), -7*f(t) + 9*k(t)),
 Eq(Derivative(k(t), t), -7*g(t) - 9*h(t))]

In [197]: sol = dsolve(eqs)                                                                                                                                                                                 

In [198]: checksysodesol(eqs, sol)                                                                                                                                                                          
Out[198]: 
(False,
 [0,
  0,
  0,
  -81*sqrt(2)*C1*sqrt(9*sqrt(109) + 95)*cos(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/28 - C1*sqrt(18*sqrt(109) + 190)*cos(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/2 + C1*sqrt(190 - 18*sqrt(109))*cos(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/2 + 9*sqrt(218)*C1*sqrt(9*sqrt(109) + 95)*cos(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/28 - 81*sqrt(2)*C2*sqrt(9*sqrt(109) + 95)*sin(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/28 - C2*sqrt(18*sqrt(109) + 190)*sin(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/2 + C2*sqrt(190 - 18*sqrt(109))*sin(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/2 + 9*sqrt(218)*C2*sqrt(9*sqrt(109) + 95)*sin(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/28 - 9*sqrt(218)*C3*sqrt(95 - 9*sqrt(109))*cos(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/28 - 81*sqrt(2)*C3*sqrt(95 - 9*sqrt(109))*cos(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/28 - C3*sqrt(190 - 18*sqrt(109))*cos(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/2 + C3*sqrt(18*sqrt(109) + 190)*cos(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/2 - 9*sqrt(218)*C4*sqrt(95 - 9*sqrt(109))*sin(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/28 - 81*sqrt(2)*C4*sqrt(95 - 9*sqrt(109))*sin(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/28 - C4*sqrt(190 - 18*sqrt(109))*sin(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/2 + C4*sqrt(18*sqrt(109) + 190)*sin(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/2])
```

When you are free, maybe you can take a look at whats going wrong. Sidenote: I checked thoroughly and I think the main problem is probably with matrix exponential.

oscarbenjamin commented:
I think that the answer is correct in that case. Probably checkodesol needs to be improved:
```julia
In [80]: eqs1 = [Eq(Derivative(f(t), t), h(t)), 
    ...:  Eq(Derivative(g(t), t), k(t)), 
    ...:  Eq(Derivative(h(t), t), -7*f(t) + 9*k(t)), 
    ...:  Eq(Derivative(k(t), t), -7*g(t) - 9*h(t))]                                                                              

In [81]: sol1 = dsolve(eqs1)                                                                                                      

In [82]: checkodesol(eqs1, sol1)                                                                                                  
Out[82]: 
(False,
 [0,
  0,
  0,
  -81*sqrt(2)*C1*sqrt(9*sqrt(109) + 95)*cos(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/28 - C1*sqrt(18*sqrt(109) + 190)*cos(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/2 + C1*sqrt(190 - 18*sqrt(109))*cos(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/2 + 9*sqrt(218)*C1*sqrt(9*sqrt(109) + 95)*cos(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/28 - 81*sqrt(2)*C2*sqrt(9*sqrt(109) + 95)*sin(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/28 - C2*sqrt(18*sqrt(109) + 190)*sin(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/2 + C2*sqrt(190 - 18*sqrt(109))*sin(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/2 + 9*sqrt(218)*C2*sqrt(9*sqrt(109) + 95)*sin(sqrt(2)*t*sqrt(9*sqrt(109) + 95)/2)/28 - 9*sqrt(218)*C3*sqrt(95 - 9*sqrt(109))*cos(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/28 - 81*sqrt(2)*C3*sqrt(95 - 9*sqrt(109))*cos(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/28 - C3*sqrt(190 - 18*sqrt(109))*cos(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/2 + C3*sqrt(18*sqrt(109) + 190)*cos(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/2 - 9*sqrt(218)*C4*sqrt(95 - 9*sqrt(109))*sin(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/28 - 81*sqrt(2)*C4*sqrt(95 - 9*sqrt(109))*sin(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/28 - C4*sqrt(190 - 18*sqrt(109))*sin(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/2 + C4*sqrt(18*sqrt(109) + 190)*sin(sqrt(2)*t*sqrt(95 - 9*sqrt(109))/2)/2])

In [83]: res = _[1][3]                                                                                                            

In [85]: sqrtdenest(res).simplify()                                                                                               
Out[85]: 0
```
Also the matrix exponential can be improved to give a consistent result regardless of the ordering.

oscarbenjamin commented on sympy/solvers/ode/tests/test_systems.py:
```diff
@@ -1922,3 +2018,77 @@ def test_dsolve():
+    (log(t)+t**2)*2*diff(x(t),t)+(log(t)+t**2)*9*diff(y(t),t)))
+    _x1 = log(t)**2
+    _x2 = sqrt(22)
+    _x3 = sqrt(9*_x1 + t**4 + 6*t**2*log(t) - 6*t**2 - 18*log(t) + 9)
```
We need to find a better way of handling examples like this. Rather than using the simpsol function to create the code for enormous complicated test solutions we should use it to study the structure of the solutions and how to make them simpler. If they can't be made simpler then we need tests that give simpler answers because these are just unmanageable.

If someone changed something somewhere and it made a minus sign or something appear somewhere in this solution then they would need to regenerate this test and they would find it very difficult to do that. This needs to be written in a way that is understandable and updateable.

mijo2 replied:
So, should we remove these type of test cases which were made simpler using your simplification method?

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -1236,6 +1236,28 @@ def _component_solver(eqs, funcs, t):
+        new_eqs = []
+
+        for i in range(1, sys_order[func]):
+            new_func = Function(Dummy())(t)
```
We should give these dummies names so e.g. `x(t).diff(t)` would be `Function(Dummy('x_1'))(t)` and `x(t)` would be `Function(Dummy('x_0')(t))`.

mijo2 replied:
I would do that for functions introduced but should I do it for dependent variables like `x(t)` since we need to access the solutions later using a dictionary and the original dependent variables don't really have to be renamed.

oscarbenjamin commented:
I guess that there needs to be a specific handler for Euler ODEs in the higher order case: https://docs.sympy.org/latest/modules/solvers/ode.html#linear-2-equations-order-2-type-9

That can be generalised to arbitrary order and an arbitrary number of equations.

mijo2 commented:
> Also the matrix exponential can be improved to give a consistent result regardless of the ordering.

So, should I add the `sqrtdenest` method in `checkodesol` to better check the solutions?

> I guess that there needs to be a specific handler for Euler ODEs in the higher order case: 

Not only that, I think that linear 2 equations, order 2, type 5-11 aren't solvable by our method. I have done the analysis and found this out. Results are given [here](https://docs.google.com/document/d/19etXqk0ckHGf7qNo1ClAI8TfmgOWTxAtBoeOuyPOUU8/edit?usp=sharing)

mijo2 committed:
solvers.ode: Changed the dummy variables naming

mijo2 committed:
solvers.ode: Added sqrtdenest simplification in checksysodesol

For a particular test case, the solution was correct but due to
the inability of checksysodesol to correctly evaluate an expression,
the solution was shown as incorrect. This was fixed in this PR.

mijo2 committed:
solvers.ode: Fixed the ordering of _extract_funcs

When there are multiple number of dependent variables in a particular
equation of a system, then there is a chance that the order of dependent
variables in the `funcs` variable is different depending on the Python
version. This was fixed in this commit.

mijo2 commented:
@oscarbenjamin Is there is a sympy function that can collect on sums? For example:

```
>>> e = t*Derivative(x(t), t) - x(t)
>>> collect_sum(b*t*Derivative(x(t), t) - b*x(t) + t**2, e)
b*(t*Derivative(x(t), t) - x(t)) + t**2
```

I tried finding such function but nothing worked like above.

mijo2 committed:
solvers.ode: Added a unique substitution method to reduce a system

Substitution method to reduce a second order system into a first order
one. The function isn't complete yet.

oscarbenjamin commented:
> Is there is a sympy function that can collect on sums?

Not that I know of but you could check the coefficient of each term in the sum:
```julia
In [17]: [expr.coeff(a) for a in Add.make_args(e)]
Out[17]: [b, b]
```
Since the coefficients are the same we know that we can construct the expression as
```julia
In [26]: expr = b*t*Derivative(x(t), t) - b*x(t) + t**2

In [27]: e = t*Derivative(x(t), t) - x(t)

In [28]: terms = Add.make_args(e)

In [29]: coeffs = [expr.coeff(a) for a in terms]

In [30]: len(set(coeffs)) == 1
Out[30]: True

In [31]: coeffs[0] * e + (expr - coeffs[0] * e).expand()
Out[31]: 
  ⎛  d              ⎞    2
b⋅⎜t⋅──(x(t)) - x(t)⎟ + t 
  ⎝  dt             ⎠ 
```
Does that help? It only works when collecting a linear combination. That could be a useful function to be added somewhere. If it needs to work for higher powers of `e` then that could be quite complicated.

Another possibility is to use `horner` if you have an idea what a coefficient might be:
```julia
In [38]: horner(expr, b)                                                                                                                       
Out[38]: 
  ⎛  d              ⎞    2
b⋅⎜t⋅──(x(t)) - x(t)⎟ + t 
  ⎝  dt             ⎠
```
I guess in general it would be nice to have something like `factor_terms` that can identify common factors in a subset of the terms of an Add.

mijo2 committed:
solvers.ode: Added second order substitution type 1

Added the function call in dsolve_system

mijo2 committed:
solvers.ode: Added _second_order_to_first_order function

This function specially deals with second order systems where
particular substitutions can be made to reduce the system.

mijo2 commented:
> Does that help?

It did help but turns out there was a much simpler way to check if the substitution can be done. We could have just taken the matrics `A1` and `A0`:
```
(A2, A1, A0), b = linear_ode_to_matrix(eqs, funcs, t, 2)
```
and checked if `A1 + A0*t == 0`. This check will work seeing the structure of the system and thus no complex collections were needed. Along with that, we can just use `A1` or `A0*t` and `b` to get the reduced system.

oscarbenjamin commented:
> turns out there was a much simpler way

Good work!

mijo2 committed:
solvers.ode: Removed the old solvers and moved their test cases

Removed solvers Linear, 2 Equations, Order 2, Type 5, 8, 11

oscarbenjamin commented:
I suppose that Cauchy-Euler systems can be checked for from the matrices in a similar way.

mijo2 commented:
@oscarbenjamin Yes, we just need to get all the lower order matrices and check by dividing with the appropriate power of `t` if the coefficient matrix is constant or not and also check if the non homogeneous term is 0 or not.

mijo2 commented:
One more thing to note is that for now, the substitution method is adjusted so that, it wouldn't probably be compatible with the system reduction with weakly and strongly connected components. I would improve the implementation of these techniques when they have been thoroughly tested and this would maybe be the final things to be done in this PR.

mijo2 committed:
solvers.ode: Added the higher order reduction in solvers

The higher order reduction usage was removed from dsoe_system and
added to the ode solvers for better code quality and more coverage
on the number of cases solved.

mijo2 commented:
@oscarbenjamin I was wondering about how we would detect that a matrix which is of this form:
```
M(t) = expr(t) * A
```
Where `expr(t)` is a scalar and `A` is a constant coefficient matrix. So, I found a method to do that. It is easy to observe that if a matrix is of the above form, then its det will be of the form below:
```
|M(t)| = c * expr(t)**2
```
Where `c` is constant. So, to check if a matrix `M(t)` is of the form that we require it to be, we just have to check if `M(t) / sqrt(|M(t)|)` is a constant matrix or not. 

Does this technique sound correct? I tried to check if this technique may detect other matrices as `True` when they shouldn't but I checked with a rudimentary proof and it works. I still want to confirm this. For now, I will implement the type10 solver method with this logic to detect the system.

oscarbenjamin commented:
Multiplying an NxN matrix by a scalar `b` scales the determinant by `b**N` so
```
| b B | = b^N | B |
```
One problem with this method would be if the determinant is zero. Another is that computing the determinant is unnecessarily expensive for this.

You just need to find a single non-zero element and divide the rest of the matrix by that element. So if you find a non-zero element that is `2*exp(t)` then you can separate the part involving `t` multiplicatively like:
```julia
In [4]: (2*exp(t)).as_independent(t, as_Add=False)                                                                                             
Out[4]: 
⎛    t⎞
⎝2, ℯ ⎠
```
Then you divide the whole matrix by `exp(t)` and if that gives a constant matrix then you have the required form and you have identified the function `f(t)`.

mijo2 commented:
> Multiplying an NxN matrix by a scalar b scales the determinant by b**N so

Ya, sorry for the rookie mistake.

> Then you divide the whole matrix

Thanks for the much better approach, will start working on it.

mijo2 committed:
solvers.ode: Added the general version of Linear, Order 2, Type 10

mijo2 committed:
solvers.ode: Added general case Euler systems solver

mijo2 committed:
solvers.ode: Added Linear n equations, Order 1, Type 5 and 6

mijo2 commented on sympy/solvers/ode/systems.py:
```diff
@@ -160,7 +174,35 @@ def linodesolve_type(A, t, b=None):
+        F_t = integrate(factor_terms[0], t)
+        inverse = solveset(Eq(t_, F_t), t)
+
+        # Note: A simple way to check if a function is invertible
```
@oscarbenjamin Does this method seem correct?

oscarbenjamin replied:
Yes, although ideally there should be a function for computing the inverse of a sympy expression.

mijo2 replied:
I was also thinking of filtering inverses which has `Piecewise` operations on it to avoid overcomplicating the solutions when it doesn't have to be.

oscarbenjamin commented:
What is the plan for this PR?

Can you update the OP to explain what methods are being removed and what cases are handled by the new code?

As GSOC is nearly finished I think we need to get this merged even if that leaves more methods to be added later.

mijo2 commented:
I will update the OP and wrap up this PR in a day or two. 

Along with that, we can work on simplifications before wrapping it up and I was also thinking of listing out the methods that we weren't able to implement(mainly because the higher order reduction was way bigger than previously thought) in the tracker issue, so that someone else can take it up in the future. 

I won't be able to work or contribute immediately after gsoc since I have gotten busy with some other personal commitments.

mijo2 commented:
Can we remove Linear, 2 equations, Order 2, Type 3? There are no equations describing the form and no test cases but I think it can be solved by our method. I just want to confirm before removing the system.

oscarbenjamin commented:
The type3 solve matches ODEs of the form:
```
a1 x'' + a y' + b x
a2 y'' - a x' + b x
```
Here's an example:
```julia
In [1]: x, y = symbols('x, y', cls=Function)                                                                                                                  

In [2]: eq = [Eq(x(t).diff(t, 2) + y(t).diff(t) + x(t), 0), Eq(y(t).diff(t, 2) - x(t).diff(t) + y(t), 0)]                                                     

In [3]: from sympy.solvers.ode.ode import classify_sysode                                                                                                     

In [4]: classify_sysode(eq)                                                                                                                                   
Out[4]: 
{'no_of_equation': 2,
 'eq': [x(t) + Derivative(x(t), (t, 2)) + Derivative(y(t), t),
  y(t) - Derivative(x(t), t) + Derivative(y(t), (t, 2))],
 'func': [x(t), y(t)],
 'order': {x(t): 2, y(t): 2},
 'func_coeff': {(0, x(t), 0): 1,
  (0, x(t), 1): 0,
  (0, x(t), 2): 1,
  (0, y(t), 0): 0,
  (0, y(t), 1): 1,
  (0, y(t), 2): 0,
  (1, x(t), 0): 0,
  (1, x(t), 1): -1,
  (1, x(t), 2): 0,
  (1, y(t), 0): 1,
  (1, y(t), 1): 0,
  (1, y(t), 2): 1},
 'is_linear': True,
 'type_of_equation': 'type3'}

In [5]: dsolve(eq)                                                                                                                                            
Out[5]: 
⎡             ⎛  ⎛  1   √5⎞⎞         ⎛  ⎛  1   √5⎞⎞         ⎛   ⎛  √5   1⎞⎞         ⎛   ⎛  √5   1⎞⎞                 ⎛  ⎛  1   √5⎞⎞         ⎛  ⎛  1   √5⎞⎞   
⎢x(t) = C₁⋅cos⎜t⋅⎜- ─ + ──⎟⎟ + C₂⋅sin⎜t⋅⎜- ─ + ──⎟⎟ + C₃⋅cos⎜-t⋅⎜- ── - ─⎟⎟ - C₄⋅sin⎜-t⋅⎜- ── - ─⎟⎟, y(t) = - C₁⋅sin⎜t⋅⎜- ─ + ──⎟⎟ + C₂⋅cos⎜t⋅⎜- ─ + ──⎟⎟ + 
⎣             ⎝  ⎝  2   2 ⎠⎠         ⎝  ⎝  2   2 ⎠⎠         ⎝   ⎝  2    2⎠⎠         ⎝   ⎝  2    2⎠⎠                 ⎝  ⎝  2   2 ⎠⎠         ⎝  ⎝  2   2 ⎠⎠   

      ⎛   ⎛  √5   1⎞⎞         ⎛   ⎛  √5   1⎞⎞⎤
C₃⋅sin⎜-t⋅⎜- ── - ─⎟⎟ + C₄⋅cos⎜-t⋅⎜- ── - ─⎟⎟⎥
      ⎝   ⎝  2    2⎠⎠         ⎝   ⎝  2    2⎠⎠⎦

In [6]: sol = dsolve(eq)                                                                                                                                      

In [7]: checkodesol(eq, sol)                                                                                                                                  
Out[7]: (True, [0, 0])

In [8]: eq                                                                                                                                                    
Out[8]: 
⎡         2                                           2          ⎤
⎢        d          d                    d           d           ⎥
⎢x(t) + ───(x(t)) + ──(y(t)) = 0, y(t) - ──(x(t)) + ───(y(t)) = 0⎥
⎢         2         dt                   dt           2          ⎥
⎣       dt                                          dt           ⎦
```
A test should be added for that case. The solver can be removed if all ODE systems of that type are now handled by rewriting as 1st order.

The type3 solver is buggy:
```julia
In [9]: eq = [Eq(x(t).diff(t, 2) + y(t).diff(t) - x(t), 0), Eq(y(t).diff(t, 2) - x(t).diff(t) - y(t), 0)]                                                     

In [10]: dsolve(eq)                                                                                                                                           
---------------------------------------------------------------------------
UnboundLocalError 

In [11]: a, b = symbols('a, b')                                                                                                                               

In [12]: eq = [Eq(x(t).diff(t, 2) + a*y(t).diff(t) + b*x(t), 0), Eq(y(t).diff(t, 2) - a*x(t).diff(t) + b*y(t), 0)]                                            

In [13]: dsolve(eq)                                                                                                                                           
---------------------------------------------------------------------------
TypeError

In [14]: eq = [Eq(2*x(t).diff(t, 2) + y(t).diff(t) + x(t), 0), Eq(y(t).diff(t, 2) - x(t).diff(t) + y(t), 0)]                                                  

In [15]: classify_sysode(eq)                                                                                                                                  
Out[15]: 
{'no_of_equation': 2,
 'eq': [x(t) + 2*Derivative(x(t), (t, 2)) + Derivative(y(t), t),
  y(t) - Derivative(x(t), t) + Derivative(y(t), (t, 2))],
 'func': [x(t), y(t)],
 'order': {x(t): 2, y(t): 2},
 'func_coeff': {(0, x(t), 0): 1,
  (0, x(t), 1): 0,
  (0, x(t), 2): 2,
  (0, y(t), 0): 0,
  (0, y(t), 1): 1,
  (0, y(t), 2): 0,
  (1, x(t), 0): 0,
  (1, x(t), 1): -1,
  (1, x(t), 2): 0,
  (1, y(t), 0): 1,
  (1, y(t), 1): 0,
  (1, y(t), 2): 1},
 'is_linear': True,
 'type_of_equation': 'type3'}

In [16]: dsolve(eq)                                                                                                                                           
Out[16]: 
⎡             ⎛t⎞         ⎛t⎞                                         ⎛t⎞         ⎛t⎞                        ⎤
⎢x(t) = C₁⋅cos⎜─⎟ + C₂⋅sin⎜─⎟ + C₃⋅cos(t) - C₄⋅sin(t), y(t) = - C₁⋅sin⎜─⎟ + C₂⋅cos⎜─⎟ + C₃⋅sin(t) + C₄⋅cos(t)⎥
⎣             ⎝2⎠         ⎝2⎠                                         ⎝2⎠         ⎝2⎠                        ⎦

In [17]: sol = dsolve(eq)                                                                                                                                     

In [18]: checkodesol(eq, sol)                                                                                                                                 
Out[18]: (False, [0, -C1*sin(t/2)/4 + C2*cos(t/2)/4 + C3*sin(t) + C4*cos(t)])
```
If those are all fixed now then regression tests can be added.

mijo2 commented:
> If those are all fixed now then regression tests can be added.

Sure, would add the test cases.

Along with that, I have concerns regarding `type6` solver. Namely, taking the below system as an example:
```
In [9]: eqs1                                                                                                                                                                                                
Out[9]: 
⎡d              2⋅(f(x) - g(x))  d          -2⋅(-f(x) + g(x)) ⎤
⎢──(f(x)) = 1 + ───────────────, ──(g(x)) = ──────────────────⎥
⎣dx                    x         dx                 x         ⎦

In [7]: dsolve_system(eqs1)                                                                                                                                                                                 
Out[7]: 
[[Eq(f(x), 2*C1*log(x) + C1 + 2*C2 + 2*log(x)*Integral(exp(x_), (x_, log(x))) + 2*Integral(-x_*exp(x_), (x_, log(x))) + Integral(exp(x_), (x_, log(x)))),
  Eq(g(x), 2*C1*log(x) + 2*C2 + 2*log(x)*Integral(exp(x_), (x_, log(x))) + 2*Integral(-x_*exp(x_), (x_, log(x))))]]

```

As you can see, we have the internally introduced independent variable present here. What should we do with regard to this? This doesn't happen with `type5` but with `type6` also, if we later have the doit method as `True` by default, it won't happen but what should we do about the case when `doit` is `False`?

mijo2 committed:
solvers.ode: Added test cases for type 5 and type 6

mijo2 committed:
solvers.ode: Added dsolve call for solving one equation systems

mijo2 committed:
solvers.ode: Added regression test case for issue #19238

mijo2 committed:
solvers.ode: Removed Linear, 2 Equations, Order 2, Type 3 solver

oscarbenjamin commented:
> we have the internally introduced independent variable present here

It will also happen if the integral can't be evaluated. I think it's fine but we should try to prevent the symbol name from clashing e.g. by using `uniequly_named_symbol`. That can be done later though (if it isn't already).

mijo2 commented:
> It will also happen if the integral can't be evaluated. I think it's fine but we should try to prevent the symbol name from clashing e.g. by using uniequly_named_symbol. That can be done later though (if it isn't already).

Ok sure thing. 

Also, I have updated the description to be more detailed because many techniques were added in this PR.

mijo2 commented:
I think this PR can be merged after all the tests have passed. You can have a look for review and maybe we can get this merged today if there aren't any big issues with the code.

oscarbenjamin commented:
The tests have failed.

oscarbenjamin commented on doc/src/modules/solvers/ode.rst:
```diff
@@ -221,50 +221,10 @@ Linear, 2 equations, Order 1, Type 7
-
 Linear, 2 equations, Order 2, Type 7
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 .. autofunction:: sympy.solvers.ode.ode::_linear_2eq_order2_type7
```
Why is type7 still here?

mijo2 replied:
I just noticed that the type7 solver can be removed now. Earlier, I don't know why the systems of this type weren't getting solved. I would remove this solver right away, but we still need to implement a method to reduce the systems which have their minium order term greater than 0.

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -75,10 +79,17 @@ def linodesolve_type(A, t, b=None):
+
     If the system is non-constant coefficient homogeneous, then "type3" is returned
+
     If the system is non-constnt coefficient non-homogeneous, then "type4" is returned
```
`constnt -> constant`

oscarbenjamin commented:
The structure of the solvers here and all of the types and the different functions seems somewhat convoluted to me. Given that all of the specific solvers here are only applicable to linear systems it would make more sense for them to have a standard structure for the arguments that are being passed around as a representation of the system. The obvious form would be the one that is returned from `linear_ode_to_matrix` (a list of matrices and rhs). I would have expected that `linodesolve` would take that form as its arguments. The other types and solvers are mostly ways of transforming that system so they could take and return a list of matrices along with some way to represent transforming the solution back to the original variables. I think it probably makes more sense to think in terms of transformations that can be applied rather than "types" of system in general and it would be good to eliminate the whole classify-then-solve approach eventually.

There are however lots of good improvements in functionality here. This greatly extends the capabilities of dsolve for systems so we should get this merged now and we can defer structural improvements until later when we can eliminate all of the old dsolve for systems code.

I think we can merge this if it's fixed so that the tests pass and then focus on simplification of solutions as that's the most important thing to finish before the end of GSOC.

mijo2 commented on sympy/solvers/ode/systems.py:
```diff
@@ -731,19 +814,37 @@ def linodesolve(A, t, b=None, B=None, type="auto", doit=False):
+                A = system_info['A']
+                b = system_info['b']
+
+    if any(type == typ for typ in ["type1", "type2", "type5", "type6"]):
```
yes, I would write in this format in the next commit.

mijo2 commented:
>  I would have expected that linodesolve would take that form as its arguments.

Yeah, if there is time left after simplification PR, we can maybe focus on bringing this together.

mijo2 committed:
solvers.ode: Removed Order 2 Type 7 solver

mijo2 committed:
Merge branch 'master' into sysorder

mijo2 committed:
solvers.ode: Updated a test case

mijo2 committed:
solvers.ode: Added a new solver

Added a higher order solver that can solve systems of type 6 order 2
Its an extension of that type.

mijo2 commented:
The checks have passed.

oscarbenjamin commented on sympy/solvers/ode/ode.py:
```diff
@@ -1957,9 +1956,8 @@ def linearity_check(eqs, j, func, is_linear_):
                 if order_eq == 1:
                     type_of_equation = check_linear_2eq_order1(eq, funcs, func_coef)
-                elif order_eq == 2:
-                    type_of_equation = check_linear_2eq_order2(eq, funcs, func_coef)
```
Why is `check_linear_2eq_order1` still needed?

mijo2 replied:
For linear 2 equations order 1 type 6 and type 7 I think. There were some test cases solved by them. One of them is in the doctests of `dsolve`

oscarbenjamin replied:
Is type6 not handled by the current solvers? Does type7 actually work? I thought it just produced the bogus solutions with `x0(t)` in them...

mijo2 replied:
I haven't checked for the validity of those solvers yet.

oscarbenjamin replied:
That's fine. Let's focus on simplification now

oscarbenjamin commented:
This... looks good. Let's put this one to rest so you can finish up.

Good work @mijo2 !
